src/core/public/http/fetch.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Url from 'url';
+
+import { merge } from '../../utils';
+import { IBasePath, HttpInterceptor, HttpHandler, HttpFetchOptions, IHttpResponse } from './types';
+import { HttpFetchError } from './http_fetch_error';
+import { HttpInterceptController } from './http_intercept_controller';
+import { HttpResponse } from './response';
+import { interceptRequest, interceptResponse } from './intercept';
+import { HttpInterceptHaltError } from './http_intercept_halt_error';
+
+interface Params {
+  basePath: IBasePath;
+  kibanaVersion: string;
+}
+
+const JSON_CONTENT = /^(application\/(json|x-javascript)|text\/(x-)?javascript|x-json)(;.*)?$/;
+const NDJSON_CONTENT = /^(application\/ndjson)(;.*)?$/;
+
+export class FetchService {
+  private readonly interceptors = new Set<HttpInterceptor>();
+
+  constructor(private readonly params: Params) {}
+
+  public intercept(interceptor: HttpInterceptor) {
+    this.interceptors.add(interceptor);
+    return () => this.interceptors.delete(interceptor);
+  }
```

Comment:
Agreed. This is named as such to be compatible with the `ui/kfetch` module as a drop in replacement. However, I think we could change this one easily since it's only used by a handful of plugins. Let me do that in a separate PR to keep the potential impact of this initial refactor minimal.